### PR TITLE
dyno: use chplenv to inform `testInteractive`

### PR DESCRIPTION
**Pull request by DanilaFe**
This PR extracts some of the command-running functionality
into Dyno library code, and uses that functionality to implement
`getChplEnv`. By reading environment variables, code that
uses Dyno can properly configure search paths. In this
PR, `testInteractive` is modified to properly find generated
system modules on systems other than `x86-gnu`. I expect
that a similar approach will be used when testing ranges.

Reviewed by @mppf - thanks!

TESTING
- [x] Paratest and GitHub CI.

### CODE REVIEW

File: compiler/dyno/include/chpl/framework/Context.h [hunk 1]
```diff
@@ -247,6 +255,22 @@ class Context {
   Context();
   ~Context();
 
+  /**
+    Store a string corresponding to CHPL_HOME in the context. The context does
+    not take ownership of the string, so the argument should live at least as
+    long as the context itself.
```
Comment (mppf) [inline compiler/dyno/include/chpl/framework/Context.h:261]: Can it just be a `std::string` so we don't have to worry so much? Why wouldn't you copy it?

File: compiler/dyno/include/chpl/util/subprocess.h [hunk 2]
```diff
@@ -48,7 +52,38 @@ int executeAndWait(const std::vector<std::string>& commandVec,
...
+    the CHPL_HOME environment variable, to locate printchplenv
+
+  returns
+    an error code if something went wrong while running printchplenv,
+    or a map containing key-value pairs for each variable in
+    printchplenv's output.
+ */
+llvm::ErrorOr<ChplEnvMap>
+getChplEnv(const std::map<std::string, const char*>& varMap,
+           const char* chplHome);
```
Comment (mppf) [inline compiler/dyno/include/chpl/util/subprocess.h:85]: Hmm... this one (and the related `include <unordered_map>` etc) seem unrelated to subprocesses in general. Can we make a new header like maybe `chplenv.h` or something?

I thought for a minute that you could make it a `static` helper method inside of `Context.cpp`, but that seems harder with the production compiler calling it.

File: compiler/dyno/include/chpl/framework/Context.h [hunk 3]
```diff
@@ -247,6 +255,22 @@ class Context {
...
+    Changing chplHome invalidates previous caches of printchplenv output.
+  */
+  void setChplHome(const char* chplHome);
+
+  /**
+    Run printchplenv, or return a cached result of doing so. To get output,
+    CHPL_HOME must have been provided via setChplHome; otherwise, the resulting
+    map will be empty.
+  */
+  llvm::ErrorOr<const ChplEnvMap&> getChplEnv();
```
Comment (mppf) [inline compiler/dyno/include/chpl/framework/Context.h:272]: Can you add a comment about how this will handle arguments for adjusting various settings, like `CHPL_COMM=gasnet` for `--comm=gasnet`?

At first I thought they would be arguments here, but instead I think they need to behave like `setChplHome` -- meaning they would be set on the Context.

File: compiler/dyno/include/chpl/framework/Context.h [hunk 4]
```diff
@@ -247,6 +255,22 @@ class Context {
...
   ~Context();
 
+  /**
+    Store a string corresponding to CHPL_HOME in the context. The context does
+    not take ownership of the string, so the argument should live at least as
+    long as the context itself.
+
+    Changing chplHome invalidates previous caches of printchplenv output.
+  */
+  void setChplHome(const char* chplHome);
```
Comment (mppf) [inline compiler/dyno/include/chpl/framework/Context.h:265]: Should this instead be something we can do when constructing the Context? That would allow you to stop worrying about it changing and invalidating the `getChplEnv` result.

**Review by mppf — APPROVED** [review]

File: compiler/dyno/include/chpl/framework/Context.h (hunk 3 above)
Comment (DanilaFe) [reply compiler/dyno/include/chpl/framework/Context.h:272]: The way that I think the production compiler currently operates is that it fetches this map and then updates keys relevant to itself. Do we want to modify keys in-place on Context's `ChplEnvMap`, or do we want to expect clients to construct a copy (yes, it could be big...) of the map and modify portions of it?

Comment (mppf) [reply compiler/dyno/include/chpl/framework/Context.h:265]: > The way that I think the production compiler currently operates is that it fetches this map and then updates keys relevant to itself

I think that's more of the exception than the rule, but I could be wrong. I don't think we should allow in-place mutation on Context's ChplEnvMap. If having the production compiler copy ChplEnvMap helps you implement it, that works for me. The main thing is we don't want to run `printchplenv` twice.

Comment (DanilaFe) [reply compiler/dyno/include/chpl/framework/Context.h:265]: Just one last thing: production compiler specifies overrides using a `map<string, const char*>`, but we return the mapping as `unordered_map<string, string>`. Should these be the same? I feel like for Dyno as a library, it makes sense to return strings (and thus explicitly pass ownership without leaking), but it also makes sense for the compiler to pass in a map with `const char*` keys, I guess. 

One wild idea is to make `getChplEnv` in `chplenv.h` be a template function over the input map type...

Comment (DanilaFe) [reply compiler/dyno/include/chpl/framework/Context.h:265]: Here's what that might look like: https://github.com/DanilaFe/chapel/commit/2dbdc39f23198f7ab3ef6c0675d197eb01b77849

Comment (DanilaFe) [reply compiler/dyno/include/chpl/framework/Context.h:268]: I like the 2-overload approach and a private template implementation in chplenv.cpp. I added this along with a new constructor argument for overriding initial chplevn variables in Context.
